### PR TITLE
chore:  Refactor OVHCloudSDConfig

### LINKED ISSUE #7206: Refactor `OVHCloudSDConfig`
### Component(s)

_No response_

### What is missing? Please describe.

We want to refactor the ScrapeConfig API, at least for all the tier-1 and tier-2 Service Discoveries mentioned in https://github.com/prometheus-operator/prometheus-operator/pull/6634. Currently all of tier-1 stands refactored and some from the tier-2 are left incomplete and we want to work on them.

This API has been mostly refactored, the main requirements are:
Test cases for all the API behavior we expect needs to be written.
If a scrapeconfig is specified with fields that are unsupported for the given Prometheus version in the cluster, it needs to be rejected at resource_selector.go
Keep an eye on the minimum Prometheus version required to configure certain fields, you will have to comment and also add test cases to ensure that the unsupported field is dealt with properly.

You will find that most (hopefully all) of the above rules are followed in prior refactoring PRs. You may take inspiration from the PRs that refactored different SDs before:
https://github.com/prometheus-operator/prometheus-operator/pull/6869
https://github.com/prometheus-operator/prometheus-operator/pull/6800
https://github.com/prometheus-operator/prometheus-operator/pull/6744

Especially test functions may be daunting to write at first, it's really useful to take reference from prior PRs.
I tried to cover as much as I can. This might not be a good first issue in the truest sense, but it is definitely something that is exciting to do as a newcomer. If you have any doubts, you can reach out to me or any of the other maintainers.

### Describe alternatives you've considered.

_No response_

## Environment
Kubernetes Version:
Prometheus-Operator Version:

### PR DESCRIPTION
**Pull request by slashexx**
## Description

This pull request refactors the validation logic for the OVHCloudSDConfig struct in the Prometheus Operator. The changes include improving the handling of configuration fields such as ApplicationKey, Service, Endpoint, and RefreshInterval. 

Fixes #7206 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [x] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [x] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Refactor validation and enhance test coverage for OVHCloudSDConfig in Prometheus Operator.

```

### CODE REVIEW

**Comment by slashexx** [comment]
@simonpasquier, I ran `make --always-make format generate` and these are the last few lines of the console

```
GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags="-s -X github.com/prometheus/common/version.Revision=57203f23c -X github.com/prometheus/common/version.BuildUser=slashex -X github.com/prometheus/common/version.BuildDate=20241224-20:26:00 -X github.com/prometheus/common/version.Branch=slashexx/refactor-ovhcloud-sd -X github.com/prometheus/common/version.Version=0.79.2" -o operator cmd/operator/main.go
/home/slashex/Documents/prometheus-operator/tmp/bin/mdox fmt Documentation/operator.md
/home/slashex/Documents/prometheus-operator/tmp/bin/mdox fmt Documentation/compatibility.md
1.23
sed -i.bak "s/ARG GOLANG_BUILDER=.*/ARG GOLANG_BUILDER=1.23/" \
	Dockerfile && rm Dockerfile.bak
sed -i.bak "s/ARG GOLANG_BUILDER=.*/ARG GOLANG_BUILDER=1.23/" \
	cmd/prometheus-config-reloader/Dockerfile && rm cmd/prometheus-config-reloader/Dockerfile.bak
sed -i.bak "s/ARG GOLANG_BUILDER=.*/ARG GOLANG_BUILDER=1.23/" \
	cmd/admission-webhook/Dockerfile && rm cmd/admission-webhook/Dockerfile.bakGOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags="-s -X github.com/prometheus/common/version.Revision=57203f23c -X github.com/prometheus/common/version.BuildUser=slashex -X github.com/prometheus/common/version.BuildDate=20241224-20:26:00 -X github.com/prometheus/common/version.Branch=slashexx/refactor-ovhcloud-sd -X github.com/prometheus/common/version.Version=0.79.2" -o operator cmd/operator/main.go
/home/slashex/Documents/prometheus-operator/tmp/bin/mdox fmt Documentation/operator.md
/home/slashex/Documents/prometheus-operator/tmp/bin/mdox fmt Documentation/compatibility.md
1.23
sed -i.bak "s/ARG GOLANG_BUILDER=.*/ARG GOLANG_BUILDER=1.23/" \
	Dockerfile && rm Dockerfile.bak
sed -i.bak "s/ARG GOLANG_BUILDER=.*/ARG GOLANG_BUILDER=1.23/" \
	cmd/prometheus-config-reloader/Dockerfile && rm cmd/prometheus-config-reloader/Dockerfile.bak
sed -i.bak "s/ARG GOLANG_BUILDER=.*/ARG GOLANG_BUILDER=1.23/" \
	cmd/admission-webhook/Dockerfile && rm cmd/admission-webhook/Dockerfile.bak
```

I still don't see any other modified files in this PR ? Is there any issue or is this how it's supposed to be ?

Also, are there any configuration fields like - Auth, OAuth2, BasicAuth, EnableHTTP, AllowRedirects  as a part of OVHCloudSD that need to be added ?

**Comment by mviswanathsai** [comment]
You can look at what fields are present here: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#ovhcloud_sd_config.
Also, you can take a look at https://github.com/prometheus/prometheus/blob/main/CHANGELOG.md and https://github.com/prometheus/prometheus/blob/main/discovery/ovhcloud/ovhcloud.go

File: test/e2e/scrapeconfig_test.go [hunk 1]
```diff
@@ -2119,3 +2122,148 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
...
+	// 				ApplicationKey:    "valid-app-key",
+	// 				ApplicationSecret: v1.SecretKeySelector{Key: "valid-secret-key"},
+	// 				Service:           "VPS",
+	// 			},
+	// 		},
+	// 	},
+	// 	expectedError: true,
+	// },
+	// {
+	// 	name: "Invalid ConsumerKey with empty value",
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2229]: This test is not testing for expected behavior. Right now the test fails because, `Key` has not been defined as a required field.

@simonpasquier, I think we can keep `Key` an optional field. I reckon we can make the `LocalObjectReference` a required field though. WDYT?
cc: @slashpai

File: test/e2e/scrapeconfig_test.go [hunk 2]
```diff
@@ -2119,3 +2122,148 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
...
+					Service:           "VPS",
+					RefreshInterval:   ptr.To(monitoringv1.Duration("30s")),
+				},
+			},
+		},
+		expectedError: false,
+	},
+	// {
+	// 	// Following should technically show error when the refreshInterval has invalid value for the supported version
+	// 	name: "Invalid RefreshInterval for unsupported version",
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2201]: I don't see a minimum version requirement for the `RefreshInterval` field. Could you verify?

Also, I don't see a Prometheus version specified as a part of the struct definition. We could add that later when we need to do version based tests. But as of now in the case of OVHCloud, I don't see that requirement. 

On a side note, the object will only be rejected if you make changes at resource_selector.go. IIRC, we don't reject anything ATM.

File: test/e2e/scrapeconfig_test.go [hunk 3]
```diff
@@ -2119,3 +2122,148 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
...
+			OVHCloudSDConfigs: []monitoringv1alpha1.OVHCloudSDConfig{
+				{
+					ApplicationKey:    "valid-app-key",
+					ApplicationSecret: v1.SecretKeySelector{Key: "valid-secret-key"},
+					ConsumerKey:       v1.SecretKeySelector{Key: "valid-consumer-key"},
+				},
+			},
+		},
+		expectedError: true,
+	},
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2268]: In addition to ensuring the wrong service types are rejected, I'd also like to see dedicated test cases that ensure that we allow the right service types (VPS and DedicatedServer).

File: test/e2e/scrapeconfig_test.go [hunk 4]
```diff
@@ -2119,3 +2122,148 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
...
+var OVHCloudSDTestCases = []scrapeCRDTestCase{
+	{
+		name: "Valid OVHCloudSDConfig",
+		scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
+			OVHCloudSDConfigs: []monitoringv1alpha1.OVHCloudSDConfig{
+				{
+					ApplicationKey:    "valid-app-key",
+					ApplicationSecret: v1.SecretKeySelector{Key: "valid-secret-key"},
+					ConsumerKey:       v1.SecretKeySelector{Key: "valid-consumer-key"},
+					Service:           "VPS",
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2135]: (nit) instead of using "VPS", try using the constants defined [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go#L1065-L1072).
```suggestion
					Service:           monitoringv1alpha1.VPS,
```

Comment (mviswanathsai) [reply test/e2e/scrapeconfig_test.go:2135]: We could improve the naming of the constants too. Maybe `OVHServiceVPS` and `OVHServiceDedicatedServer`.

File: test/e2e/scrapeconfig_test.go [hunk 5]
```diff
@@ -2119,3 +2122,148 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
...
+	},
+	{
+		name: "Invalid ApplicationKey with empty value",
+		scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
+			OVHCloudSDConfigs: []monitoringv1alpha1.OVHCloudSDConfig{
+				{
+					ApplicationKey:    "",
+					ApplicationSecret: v1.SecretKeySelector{Key: "valid-secret-key"},
+					ConsumerKey:       v1.SecretKeySelector{Key: "valid-consumer-key"},
+					Service:           "VPS",
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2149]: Same comment here, and for the rest of the test cases.

File: test/e2e/scrapeconfig_test.go [hunk 6]
```diff
@@ -2119,3 +2122,148 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
 		expectedError: true,
 	},
 }
+
+var OVHCloudSDTestCases = []scrapeCRDTestCase{
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2126]: (nit) Try grouping together test cases related to a given field, to make it easier to read. For example, the test cases which cover a correctly defined Service, a wrongful Service, a missing Service etc should all be next to each other. This way we can quickly see what test cases are missing, if something is redundant etc.

Comment (mviswanathsai) [reply test/e2e/scrapeconfig_test.go:2126]: In general, 
we need test cases where the field is defined correctly, defined incorrectly (this includes wrong texts, empty texts, etc), and not defined at all.

For example, you have added a test case for empty (incorrectly defined) and correctly defined `ApplicationKey`, but seem to have missed a test case for "missing `ApplicationKey`". It would be really easy to catch if you group the tests together.

File: test/e2e/scrapeconfig_test.go [hunk 7]
```diff
@@ -2119,3 +2122,148 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
...
+	// 				ApplicationSecret: v1.SecretKeySelector{Key: "valid-secret-key"},
+	// 				ConsumerKey:       v1.SecretKeySelector{Key: ""},
+	// 				Service:           "VPS",
+	// 			},
+	// 		},
+	// 	},
+	// 	expectedError: true,
+	// },
+	{
+		name: "Invalid Service type (Another Invalid Service)",
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2243]: You may remove this test case, the one above should be good enough.

File: test/e2e/scrapeconfig_test.go [hunk 8]
```diff
@@ -2119,3 +2122,148 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
...
+	// 				ConsumerKey:       v1.SecretKeySelector{Key: "valid-consumer-key"},
+	// 				Service:           "VPS",
+	// 				RefreshInterval:   ptr.To(monitoringv1.Duration("30s")),
+	// 			},
+	// 		},
+	// 	},
+	// 	expectedError: true,
+	// },
+	// {
+	// 	name: "Missing ConsumerKey",
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2216]: Not sure why this fails. Could you re-check?

**Review by mviswanathsai — CHANGES_REQUESTED** [review]
A few comments. We need to add a few more test cases and differentiate between the test cases a bit better (for example, a test case should only deal with a single field being correctly/incorrectly defined instead of broadly "valid" configurations).

Looks good overall!

File: test/e2e/scrapeconfig_test.go (hunk 1 above)
Comment (mviswanathsai) [reply test/e2e/scrapeconfig_test.go]: And by the looks of it, that is not up to us. We are consuming K8S's API. See:  https://github.com/kubernetes/api/blob/v0.32.0/core/v1/types.go#L2430-L2438.

File: test/e2e/scrapeconfig_test.go (hunk 8 above)
Comment (slashexx) [reply test/e2e/scrapeconfig_test.go:2268]: I rewrote the testcases again and the following two failed once more :-

```go
	{
		name: "Invalid ConsumerKey with empty value",
		scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
			OVHCloudSDConfigs: []monitoringv1alpha1.OVHCloudSDConfig{
				{
					ApplicationKey:    "valid-app-key",
					ApplicationSecret: v1.SecretKeySelector{Key: "valid-secret-key"},
					ConsumerKey:       v1.SecretKeySelector{Key: ""},
					Service:           monitoringv1alpha1.OVHServiceVPS,
				},
			},
		},
		expectedError: true,
	},
	{
		name: "Missing ConsumerKey",
		scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
			OVHCloudSDConfigs: []monitoringv1alpha1.OVHCloudSDConfig{
				{
					ApplicationKey:    "valid-app-key",
					ApplicationSecret: v1.SecretKeySelector{Key: "valid-secret-key"},
					Service:           monitoringv1alpha1.OVHServiceVPS,
				},
			},
		},
		expectedError: true,
	},
```
Are the consumerKey config validations in place ?

**Comment by slashexx** [comment]
I've made most of the required changes, gentle ping, PTAL @mviswanathsai :)

File: test/e2e/scrapeconfig_test.go [hunk 9]
```diff
@@ -2119,3 +2122,204 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
...
+					ApplicationSecret: v1.SecretKeySelector{Key: "valid-secret-key"},
+					ConsumerKey:       v1.SecretKeySelector{Key: "valid-consumer-key"},
+					Service:           monitoringv1alpha1.OVHServiceVPS,
+				},
+			},
+		},
+		expectedError: true,
+	},
+	{
+		name: "Valid Service field",
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2199]: (nit) Just to make it explicit, add another test case for `OVHServiceDedicatedServer`. (I can see that you have already added it as a part of another test case.)

File: test/e2e/scrapeconfig_test.go [hunk 10]
```diff
@@ -2119,3 +2122,204 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
...
+		expectedError: false,
+	},
+	// {
+	// 	name: "Invalid ConsumerKey with empty value",
+	// 	scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
+	// 		OVHCloudSDConfigs: []monitoringv1alpha1.OVHCloudSDConfig{
+	// 			{
+	// 				ApplicationKey:    "valid-app-key",
+	// 				ApplicationSecret: v1.SecretKeySelector{Key: "valid-secret-key"},
+	// 				ConsumerKey:       v1.SecretKeySelector{Key: ""},
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2260]: IIUC, the `Key` string is allowed to be empty, i.e "". Try,
```suggestion
	// 				ConsumerKey:       v1.SecretKeySelector{},
```

File: test/e2e/scrapeconfig_test.go (hunk 8 above)
Comment (mviswanathsai) [reply test/e2e/scrapeconfig_test.go:2268]: There aren't any explicit validations at the application level. The CRD spec is expected to be generated such that the given field is required. For the CRD generation we make use of `kubebuilder`.  This might have to do with how `kubebuilder` handles `+structType:=atomic`. I am unsure about the cause of this behavior though. CC: @simonpasquier

Comment (mviswanathsai) [reply test/e2e/scrapeconfig_test.go:2268]: For now, we may have to skip this test.

File: test/e2e/scrapeconfig_test.go [hunk 11]
```diff
@@ -2119,3 +2122,204 @@ var LightSailSDTestCases = []scrapeCRDTestCase{
...
+					ConsumerKey:       v1.SecretKeySelector{Key: "valid-consumer-key"},
+					Service:           monitoringv1alpha1.OVHServiceVPS,
+					RefreshInterval:   ptr.To(monitoringv1.Duration("")),
+				},
+			},
+		},
+		expectedError: false,
+	},
+	{
+		name: "Valid RefreshInterval for supported version",
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2311]: Not sure what "supported versions" means here. We can remove it.
```suggestion
		name: "Valid RefreshInterval",
```

**Review by mviswanathsai — COMMENTED** [review]
Just a few comments. I would also get some reviews from @simonpasquier and @slashpai just to be sure.

File: test/e2e/scrapeconfig_test.go (hunk 8 above)
Comment (slashexx) [reply test/e2e/scrapeconfig_test.go]: I see, I'll remove it for now then.

**Comment by slashexx** [comment]
I've made the required changes for now, @slashpai , @simonpasquier can you PTAL ?

**Comment by mviswanathsai** [comment]
LGTM! 
I would still seek clarity about the `ConsumerKey` field though. But I don't think it is anything specific to your PR.

**Review by mviswanathsai — APPROVED** [review]

**Comment by slashexx** [comment]
It's been a while since this has been open, gentle ping @slashpai , is this gtg ?

**Comment by mviswanathsai** [comment]
LGTM overall, just need a nod from @slashpai as well.

**Review by slashpai — DISMISSED** [review]
Thank you

**Comment by slashpai** [comment]
@slashexx Can you rebase your branch, changes looks good to me.

Also expect that sometimes there may be need to do multiple rebases as there are many PRs opened for scrape config refactoring in repo

**Comment by slashexx** [comment]
@slashpai The branch has been rebased !

**Review by slashpai — APPROVED** [review]
